File: backend/src/main/java/com/ecommerce/backend/service/ProductService.java
package com.ecommerce.backend.service;

import com.ecommerce.backend.dto.request.CreateProductRequest;
import com.ecommerce.backend.dto.request.UpdateProductRequest;
import com.ecommerce.backend.dto.response.ProductResponse;
import org.springframework.data.domain.Page;

import java.util.List;

public interface ProductService {

    ProductResponse createProduct(String tenantName, CreateProductRequest request);

    ProductResponse updateProduct(String tenantName, Long productId, UpdateProductRequest request);

    void deleteProduct(String tenantName, Long productId);

    ProductResponse updateStock(String tenantName, Long productId, Integer quantity);

    Page<ProductResponse> getAllProducts(String tenantName, int page, int size);

    List<ProductResponse> searchProducts(String tenantName, String keyword);

    List<ProductResponse> getProductsByCategory(String tenantName, Long categoryId);

    Page<ProductResponse> getMarketplaceProducts(int page, int size);

    ProductResponse getProductById(Long productId);
}

File: frontend/src/services/productService.js
import api from "./api";

export const getProducts = async (tenantName, page = 0, size = 10) => {
  return await api.get(`/products/${tenantName}`, {
    params: { page, size },
  });
};

export const searchProducts = async (tenantName, keyword) => {
  return await api.get(`/products/${tenantName}/search`, {
    params: { keyword },
  });
};

export const getProductsByCategory = async (tenantName, categoryId) => {
  return await api.get(`/products/${tenantName}/category/${categoryId}`);
};

export const createProduct = async (tenantName, product) => {
  return await api.post(`/products/${tenantName}`, product);
};

export const updateProduct = async (tenantName, productId, update) => {
  return await api.put(`/products/${tenantName}/${productId}`, update);
};

export const updateStock = async (tenantName, productId, quantity) => {
  return await api.patch(`/products/${tenantName}/${productId}/stock`, null, {
    params: { quantity },
  });
};

export const deleteProduct = async (tenantName, productId) => {
  return await api.delete(`/products/${tenantName}/${productId}`);
};